### DISCUSSION TITLE
Making Docker work on Windows 10

### DISCUSSION THREAD

**Discussion by vicos59** [geodesicsolutions-community/geocore-community#103] [Q&A]
@jonyo 

I was at the 5 yard line and I thought I was going to make it.... But, she blew up at step 5:

![1](https://user-images.githubusercontent.com/6634041/155756290-bdcd723a-7c5f-4315-96f0-65581a7a24ad.JPG)

This is what I see in Docker at this point:

![2](https://user-images.githubusercontent.com/6634041/155756443-8a4fdf8b-3a56-47cb-9115-5ea21fa4ea29.JPG)

![3](https://user-images.githubusercontent.com/6634041/155756470-31ba6baf-72f8-43db-be35-e86ac68c2b92.JPG)

I did 2 things differently this time:

1. Instead of relying on the git supplied by Github Desktop, I installed the git linked to VSC and used git-cli to run the install cmds
2. VSC this time prompted me to install a module to integrate with Docker, which I did.

**Comment by vicos59** [comment]
This is what I see in Docker when I try to start the container:

![4](https://user-images.githubusercontent.com/6634041/155757266-a8e49956-b6f8-4ee0-8827-8be988d3ed80.JPG)

**Comment by vicos59** [comment]
So, I switched from git-cli to powershell to run the command in step 5 and that allowed us to advance:

![5](https://user-images.githubusercontent.com/6634041/155759878-2519247e-5ebd-413f-aaa9-83e9d4b2e2f3.JPG)

Then, you get the little diddy:

`Error response from daemon: user declined directory sharing C:\Users\David\Github Projects\geodesicsolutions-community\geocore-community`

That is solved by doing this:

https://stackoverflow.com/questions/70877785/docker-error-response-from-daemon-user-declined-directory-sharing#:~:text=1%20Answer&text=You%20have%20to%20add%20C,folder%20and%20hit%20Apply%20%26%20Restart.

Then, you have winner, winner, chicken dinner!

![6](https://user-images.githubusercontent.com/6634041/155760203-09c18ba3-81ff-4306-ad93-d494814e39ee.JPG)

**Reply by vicos59** [reply]
I just noticed a Windows Alert:

![7](https://user-images.githubusercontent.com/6634041/155762695-5da9312f-028e-42b2-b80a-ddf7ebfcf4f7.JPG)

**Comment by vicos59** [comment]
From there, we have a new issue when you click 

"Open in Visual Studio Code" in Docker Desktop. It says:

`docker Error invoking remote method 'desktop-go-backend': Error: {"message":"cannot open editor: exit status 1 - "}`

I'll look for a solution.

**Comment by vicos59** [comment]
Finally, when I try to load http://localhost:8080/ (or the admin):

```
Error: Class 'Smarty' not found in /var/www/html/src/classes/php5_classes/Template.class.php:10
Stack trace:
#0 /var/www/html/src/app_top.common.php(48): require_once()
#1 [internal function]: Geo__autoload('Template')
#2 /var/www/html/src/classes/php5_classes/DataAccess.class.php(1082): spl_autoload_call('geoTemplate')
#3 /var/www/html/src/app_top.main.php(257): DataAccess->get_text(false, 59)
#4 /var/www/html/src/index.php(10): require_once('/var/www/html/s...')
#5 {main}
```

@jonyo Need some guidance from you on this. I'm not sure if it is a bug in the code or if my docker container did not install properly.

**Reply by jonyo** [reply]
So Close!!!

For that error, it means composer install did not run successfully yet...  That is the last step, you can make it run inside the container like 
```
docker-compose exec --user www-data web composer install
```

Watch for any errors.  Last time that was run I believe it failed because somehow it got some windows line endings into the `contrib/setup` file so the script was failing...  Try running it now and see if you have the same problem.

**Reply by jonyo** [reply]
I'll have to remember to look into if there is some way to make easy docker commands so you can just do something like "docker composer" or something as a shortcut...  🤔

**Reply by jonyo** [reply]
I guess there isn't really a way to do what I was thinking.

But you can set up a bash alias (just google `bash alias`) or similar.  Also you can start a console window for the `web` container, then just run `composer install` when you are inside the web container, that would do the same thing.

Let me know if you are getting error from that.

**Reply by vicos59** [reply]
Well, I encountered errors, but it looks like the website runs on locahost now. If I run 

docker-compose exec --user www-data web composer install

in Powershell, the first time I get:

```
Xdebug: [Step Debug] Time-out connecting to debugging client, waited: 200 ms. Tried: host.docker.internal:9003 (through xdebug.client_host/xdebug.client_port) :-(
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Package operations: 2 installs, 0 updates, 0 removals
Cannot create cache directory /var/www/.composer/cache/files/, or directory is not writable. Proceeding without cache
  - Downloading phpstan/phpstan (1.4.6)
  - Downloading smarty/smarty (v4.1.0)
  - Installing phpstan/phpstan (1.4.6): Extracting archive
  - Installing smarty/smarty (v4.1.0): Extracting archive
Generating autoload files
1 package you are using is looking for funding.
Use the `composer fund` command to find out more!
> bash contrib/setup
contrib/setup: line 2: $'\r': command not found
Xdebug: [Step Debug] Time-out connecting to debugging client, waited: 200 ms. Tried: host.docker.internal:9003 (through xdebug.client_host/xdebug.client_port) :-(
contrib/setup: line 4: $'\r': command not found
cp: cannot stat '/var/www/html/contrib'$'\r''/pre-commit': No such file or directory
chmod: cannot access '/var/www/html/contrib'$'\r''/../.git/hooks/pre-commit'$'\r': No such file or directory
Script bash contrib/setup handling the post-install-cmd event returned with error code 1

```

The 2nd time I try to run it:

```
Xdebug: [Step Debug] Time-out connecting to debugging client, waited: 200 ms. Tried: host.docker.internal:9003 (through xdebug.client_host/xdebug.client_port) :-(
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating autoload files
1 package you are using is looking for funding.
Use the `composer fund` command to find out more!
> bash contrib/setup
Xdebug: [Step Debug] Time-out connecting to debugging client, waited: 200 ms. Tried: host.docker.internal:9003 (through xdebug.client_host/xdebug.client_port) :-(

```

If I run composer install inside a cli launched from Docker for container geocore-community-web-1, I get:

```
# pwd
/var/www/html
# composer install
Xdebug: [Step Debug] Time-out connecting to debugging client, waited: 200 ms. Tried: host.docker.internal:9003 (through xdebug.client_host/xdebug.client_port) :-(
Installing dependencies from lock file (including require-dev)
Verifying lock file contents can be installed on current platform.
Nothing to install, update or remove
Generating autoload files
1 package you are using is looking for funding.
Use the `composer fund` command to find out more!
> bash contrib/setup
Xdebug: [Step Debug] Time-out connecting to debugging client, waited: 200 ms. Tried: host.docker.internal:9003 (through xdebug.client_host/xdebug.client_port) :-(

```

I ran dos2unix on the setup file (after I ran the command the first time). It did not appear to be in dos format to begin with. But, I do see the \r being echoed back the first time.

**Reply by vicos59** [reply]
@jonyo So, if you make a commit to the main branch, are those changes going to immediately show up in my container?

Now, I just need to get Docker to with with VSC. You would think that for programmer's tools, these things would work better out of the box.

**Reply by jonyo** [reply]
So...  The good news, the important bits worked!

Currently all that setup tries to do is set up pre-commit hook.  What that does is run a script when you try to commit something, normally to run something like a PHP sniffer to check for syntax errors and things like that.

But since the code isn't cleaned up enough to pass the checks, right now it does nothing.

Try viewing localhost:8080 or localhost:8081 (the PHP8 URL) and see if it works.

**Reply by vicos59** [reply]
Yes, the website works. I edited my last post to reflect that. Didn't see your new response.

**Reply by jonyo** [reply]
Ok I just pushed changes that disable the part that isn't actually doing anything but is giving you errors 😂 

> So, if you make a commit to the main branch, are those changes going to immediately show up in my container?

Not exactly, it isn't automagical (thankfully - you don't want it updating files on your local without you telling it to) - it only does what you tell it.  On your local, if you are on the `main` branch... And I merge changes to the main branch... Then you do a `git pull` and it will pull in those changes.

> Now, I just need to get Docker to with with VSC.

I actually... don't use the VSC plugin much for docker...  Maybe I should lol, but I don't use it so don't know how well it works.  I usually just use the docker desktop or the command line docker commands.  Old habits...

> You would think that for programmer's tools, these things would work better out of the box.

I think it may be a case that more people are on mac or ubuntu (or other linux flavors) as dev machines...  And the ones that do use windows, probably have the initial setup issues, but those are things you work out once, then after that you know what to expect so know what to do.  Or they are only issues that come up when you first set it up, so you don't see the problem after that.  Just a guess though.

**Reply by vicos59** [reply]
OK, Now that Docker is setup, I am jumping back to the Beta-4 testing thread. #95

**Comment by vicos59** [comment]
Another issue with using Docker on Windows is that it messes up existing VirtualBox machines. Now, none of my linux machines will boot. There are threads out there that say the problem is caused by settings Docker changes in regards to virtualization. Some claim it is possible to revert the settings back, at which point vBox will work and Docker will not. Then set them back and both will work. No time to mess with that these days. 

https://superuser.com/questions/1318227/why-does-ubuntu-18-04-lts-hang-at-bootup-when-installed-in-virtualbox

**Reply by jonyo** [reply]
> messes up existing VirtualBox machines

Wow I've never had that probl....  Oh yeah I'm not on Windows 10 for my dev environment 😅 ...  I did not even realize it did that on Windows.  (We used vbox in our local dev environment stack at work, but we're also using macs)

Sooo uh, have you ever considered duel booting Ubuntu desktop or something?  😂 I joke, I mean it would probably go smoother but we don't want to just say that to everyone using Windows that wants to try it.

Remember to keep track of the steps you end up doing (if you re-visit this later when you have time), to add to a wiki page to help windows users get "onboarded".

**Comment by vicos59** [comment]
> Sooo uh, have you ever considered duel booting Ubuntu desktop or something?

Well, that's why I was putting Ubuntu on vBox. I already had CentOS since my server uses it.

In any case, dual boot is not the solution for me. I work on 5 things at the same time and I need my Windows workstation up at all times.

But, yeah, I love having to reboot every time BillGatesInc installs an update on my pc.

**Comment by vicos59** [comment]
Following what VivekDev said made both Docker and vBox work together.

https://superuser.com/questions/1318227/why-does-ubuntu-18-04-lts-hang-at-bootup-when-installed-in-virtualbox

Basically disable Hyper-V, reboot, now vBox works, Docker broke, turn Hyper-V on it all works.

Only problem in my case is now all of my Docker Containers have vanished. Have to start that process all over.

**Reply by jonyo** [reply]
Ah the hyper-v thing, what an odd bug with the settings!  I wonder if that is something at the hardware level or if it is some bug in one of the respective software apps...  If the latter, probably each project thinks it's the "other" projects fault and it is hard to diagnose or something.

If you are not familiar, [here is what is up with hyper-v](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/about/) - its a thing that makes running virtual computers run faster basically.

**Reply by vicos59** [reply]
Hyper-V was off before Docker showed up in town and vBox had been running for a long time. Docker turned Hyper-V on during install and this broke vBox. Somewhere in the disable/re-enable fix, vBox realized that something was different and changed its life to adapt.

But, Docker wasn't happy that there were 2 sheriff's in town, and blew away my containers. Maybe something else I did during the 'fixin' caused that.

**Reply by vicos59** [reply]
For future reference, this is another helpful page:

https://docs.docker.com/desktop/windows/troubleshoot/#virtualization

**Comment by vicos59** [comment]
OK, the Docker setup worked flawlessly this time:

https://github.com/geodesicsolutions-community/geocore-community#using-docker

However, I recommend the following clarifications (Windows specific):

- [ ] Step 1: Remove mention of Github CLI. One of the steps bombed last time using this. Everything worked the whole way through using PowerShell run as Administrator.
- [ ] Step 1: I'm not sure it is clear to the 1st time reader that before you do anything that you need to have a copy of the GeoCoreCE repo on your local PC...regardless of how you get it there. "For the software" -- which software are we referring to? We are setting up Docker now. Oh, you mean the GeoCoreCE software. Might seem obvious, but depends who is reading. Confused me the 1st time.
- [ ] Step 4: Needs to be clear that this steps starts up the containers and as a result your Powershell window is now dedicated to that task. If you want to continue issuing command in the current PS window, Control-C then manually start the container using Docker or just start a new Powershell and get yourself to the same directory where you left off. Also, if you close the 1st PS window, the Docker container is gonna quit. It didn't detach the processes from the terminal. Bad Docker, bad!
- [ ] Step 5: "Run composer install the first time inside the container:" Just now, to me, that meant open the CLI for geocore-community-web-1 and execute `docker-compose exec --user www-data web composer install` and of course, that won't work. All they have to do is issue the command right where they left off in PS and all is well.

**Reply by jonyo** [reply]
@vicos59 Great notes!  If you have time, this can be edited in the code base, see the `README.md` in the base folder.  And if you want, there are "markdown preview" plugins available in VSC if you want to preview before you make the PR, that is usually what I do when changing the readme.

Otherwise I can make the changes when I have time.

Hmm, we might also consider mentioning the docker installation method, then link to a page in the wiki for the setup instructions, that way it is easier to edit going forward.

**Reply by jonyo** [reply]
If we go the route of making it a page in the wiki, the wiki uses the same exact markdown syntax so you should be able to just copy/paste the raw "markdown" from the readme source, into the new wiki page to create it.

**Comment by vicos59** [comment]
> If you have time

I don't see this as a high priority for anyone. Just wanted to document them here while it was fresh in my mind. Can't remember what I had for lunch yesterday.

**Comment by vicos59** [comment]
Just discovered that whichever branch you have selected in Visual Studio Code is the branch you will see viewing the docker image, i.e. http://localhost:8080.

I kept wondering why things kept breaking and fixing themselves. It's like magic!

**Comment by vicos59** [comment]
Another tale in using these dev tools on Windows. 

I pressed the button in Github Desktop to open the branch in Visual Studio Code (VSC) like I had been doing for the past 2 weeks. This time I got a popup saying it could not find it. What?! I had just closed it a bit earlier. Then I spot a small status window in the top left of the screen saying a VSC update was in process. When I clicked on that, another alert revealed itself saying the upgrade could not remove certain files; check for running processes. I did and did not see any VCS stuff. I even stopped Docker and Github. No Bueno. I had to cancel the update.

Decided to reboot to see if it would recover. Nope. Downloaded and reinstalled VSC. Luckily my settings and addons were still there. Never a dull day.
